include/aws/s3/private/s3_request_messages.h

```diff
@@ -34,50 +36,65 @@ struct aws_http_message *aws_s3_get_object_message_new(
     size_t part_size,
     bool has_range);
 
+/* Given a response body from a multipart upload, try to extract the upload id. */
+struct aws_string *aws_s3_create_multipart_upload_get_upload_id(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor *response_body);
+
+struct aws_http_message *aws_s3_abort_multipart_upload_message_new(
+    struct aws_allocator *allocator,
+    struct aws_http_message *base_message,
+    const struct aws_string *upload_id);
+
+/* TODO: maybe set a list of the headers we want */
+struct aws_http_message *aws_s3_message_util_copy_http_message(
+    struct aws_allocator *allocator,
+    struct aws_http_message *message,
+    uint32_t flags);
+
+struct aws_input_stream *aws_s3_message_util_assign_body(
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *byte_buf,
+    struct aws_http_message *out_message);
+
+AWS_EXTERN_C_BEGIN
+
 /* Create an HTTP request for an S3 Put Object request, using the original request as a basis.  Creates and assigns a
  * body stream using the passed in buffer.  If multipart is not needed, part number and upload_id can be 0 and NULL,
  * respectively. */
-struct aws_http_message *aws_s3_put_object_message_new(
+AWS_S3_API
+struct aws_http_message *aws_s3_upload_part_message_new(
     struct aws_allocator *allocator,
     struct aws_http_message *base_message,
     struct aws_byte_buf *buffer,
     uint32_t part_number,
-    const struct aws_string *upload_id);
+    const struct aws_string *upload_id,
+    bool should_compute_content_md5);
 
 /* Create an HTTP request for an S3 Create-Multipart-Upload request. */
+AWS_S3_API
 struct aws_http_message *aws_s3_create_multipart_upload_message_new(
     struct aws_allocator *allocator,
     struct aws_http_message *base_message);
 
-/* Given a response body from a multipart upload, try to extract the upload id. */
-struct aws_string *aws_s3_create_multipart_upload_get_upload_id(
-    struct aws_allocator *allocator,
-    struct aws_byte_cursor *response_body);
-
 /* Create an HTTP request for an S3 Complete-Multipart-Upload request. Creates the necessary XML payload using the
  * passed in array list of ETags.  (Each ETag is assumed to be an aws_string*)  Buffer passed in will be used to store
  * said XML payload, which will be used as the body. */
+AWS_S3_API
 struct aws_http_message *aws_s3_complete_multipart_message_new(
     struct aws_allocator *allocator,
     struct aws_http_message *base_message,
     struct aws_byte_buf *body_buffer,
     const struct aws_string *upload_id,
     const struct aws_array_list *etags);
 
-struct aws_http_message *aws_s3_abort_multipart_upload_message_new(
+/* Add content-md5 header to the http message passed in. The MD5 will be computed from the input_buf */
+AWS_S3_API
+int aws_s3_message_util_add_content_md5_header(
```

Comment:
that's the only function we add, the other changes are just moving things around, as they are used from tests